providers/profitbricks/src/main/java/org/jclouds/profitbricks/compute/extensions/ProfitBricksImageExtension.java

```diff
@@ -0,0 +1,139 @@
+   private final Function<Provisionable, Image> imageTransformer;
+   private final ListeningExecutorService userExecutor;
+
+   @Inject
+   ProfitBricksImageExtension(ProfitBricksApi api,
+           @Named(TIMEOUT_SNAPSHOT_AVAILABLE) Predicate<String> snapshotAvailablePredicate,
+           Function<Provisionable, Image> imageTransformer,
+           @Named(Constants.PROPERTY_USER_THREADS) ListeningExecutorService userExecutor) {
+      this.api = api;
+      this.snapshotAvailablePredicate = snapshotAvailablePredicate;
+      this.imageTransformer = imageTransformer;
+      this.userExecutor = userExecutor;
+   }
+
+   @Override
+   public ImageTemplate buildImageTemplateFromNode(String name, String id) {
+      Server server = api.serverApi().getServer(id);
+
+      if (server == null)
+         throw new NoSuchElementException("Cannot find server with id: " + id);
+
+      List<Storage> storages = server.storages();
+      if (storages.isEmpty() || !Iterables.tryFind(storages, matchBootDevice).isPresent())
+         throw new NoSuchElementException("Server " + id + " does not contain a boot device");
+
+      return new ImageTemplateBuilder.CloneImageTemplateBuilder().nodeId(id).name(name).build();
+   }
+
+   @Override
+   public ListenableFuture<Image> createImage(ImageTemplate template) {
+      checkState(template instanceof CloneImageTemplate, "profitbricks only supports creating images through cloning.");
+      final CloneImageTemplate cloneTemplate = (CloneImageTemplate) template;
+      String serverId = cloneTemplate.getSourceNodeId();
+
+      Server server = api.serverApi().getServer(serverId);
+      Storage bootDevice = Iterables.find(server.storages(), matchBootDevice);
+
+      final Snapshot requested = api.snapshotApi().createSnapshot(
+              Snapshot.Request.creatingBuilder()
+              .storageId(bootDevice.id())
+              .name(template.getName())
+              .description(template.getName() + " (created with jclouds)")
+              .build());
+
+      return userExecutor.submit(new Callable<Image>() {
+         @Override
+         public Image call() throws Exception {
+            if (snapshotAvailablePredicate.apply(requested.id())) {
+               Snapshot built = api.snapshotApi().getSnapshot(requested.id());
+               return imageTransformer.apply(built);
+            }
+
+            throw new UncheckedTimeoutException("Image was not created within the time limit: "
+                    + cloneTemplate.getName());
+         }
+      });
+   }
+
+   @Override
+   public boolean deleteImage(String id) {
+      Snapshot snapshot = api.snapshotApi().getSnapshot(id);
+      if (snapshot != null)
+         try {
+            logger.debug(">> deleting snapshot %s..", id);
+            return api.snapshotApi().deleteSnapshot(id);
+         } catch (Exception ex) {
+            logger.error(ex, ">> error deleting snapshot %s..", id);
+         }
+      return true;
```

Comment:
Should this better be a `return false`, to cover the case when an official Image is provided and also cover the return value with the `catch` block is executed?